src/coord/src/coord/dataflow_builder.rs

```diff
@@ -131,37 +121,29 @@ impl<'a> DataflowBuilder<'a> {
                             }
                         }
 
-                        let mut connector = source.connector.clone();
-
-                        let persist_details = self
-                            .catalog
-                            .persist()
-                            .source_persist_desc_from_serialized(
-                                &source.connector,
-                                source.persist_details.clone(),
-                            )
-                            .map_err(CoordError::Persistence)?;
-
-                        // TODO: I don't like that we're injecting this into the otherwise "pristine"
-                        // immutable SourceConnector. We should clean this up once we have an
-                        // ingestd/dataflowd split, where we probably want to send SourceConnector only to
-                        // ingestd (and always with persistence details) and dataflowd will never see the
-                        // current style of SourceConnector.
-                        match &mut connector {
-                            SourceConnector::External { persist, .. } => {
-                                assert!(persist.is_none());
-                                *persist = persist_details;
-                            }
-                            SourceConnector::Local { .. } => unreachable!(),
+                        let mut source_description =
+                            self.catalog.source_description_for(*id).unwrap();
+
+                        // Manipulate the source description, if something about persistence is true.
+                        // TODO: explain what is going on here better.
+                        if let SourceConnector::External { persist, .. } =
+                            &mut source_description.connector
+                        {
+                            assert!(persist.is_none());
+
+                            let persist_details = self
```

Comment:
@aljoscha This code is now only called in the case of an external connector, as it is the case that the results are used. This is a behavioral change in that previously it would be called for all connectors and then have the result discarded. Please lmk if this is an important behavioral change.

Reply:
I think this all gets much clearer with https://github.com/MaterializeInc/materialize/pull/10347/commits/06879b6b193b2170db40718fb7d35bf95998e5d9.

Reply:
This is correct! I'll also take a look at @benesch's PR now.